drv/stm32g0-rcc-api/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+
+//! Client API for the STM32G0 RCC server.
+
+#![no_std]
+
+use userlib::*;
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum RccError {
+    NoSuchPeripheral = 1,
+}
+
+impl From<u32> for RccError {
+    fn from(x: u32) -> Self {
+        match x {
+            1 => RccError::NoSuchPeripheral,
+            _ => panic!(),
+        }
+    }
+}
+
+impl Rcc {
+    /// Requests that the clock to a peripheral be turned on.
+    ///
+    /// This operation is idempotent and will be retried automatically should
+    /// the RCC server crash while processing it.
+    ///
+    /// # Panics
+    ///
+    /// If the RCC server has died.
+    pub fn enable_clock(&self, peripheral: Peripheral) {
+        // We are unwrapping here because the RCC server should not return
+        // NoSuchPeripheral for a valid member of the Peripheral enum.
+        self.enable_clock_raw(peripheral as u32).unwrap()
+    }
+
+    /// Requests that the clock to a peripheral be turned off.
+    ///
+    /// This operation is idempotent and will be retried automatically should
+    /// the RCC server crash while processing it.
+    ///
+    /// # Panics
+    ///
+    /// If the RCC server has died.
+    pub fn disable_clock(&self, peripheral: Peripheral) {
+        // We are unwrapping here because the RCC server should not return
+        // NoSuchPeripheral for a valid member of the Peripheral enum.
+        self.disable_clock_raw(peripheral as u32).unwrap()
+    }
+
+    /// Requests that the reset line to a peripheral be asserted.
+    ///
+    /// This operation is idempotent and will be retried automatically should
+    /// the RCC server crash while processing it.
+    ///
+    /// # Panics
+    ///
+    /// If the RCC server has died.
+    pub fn enter_reset(&self, peripheral: Peripheral) {
+        // We are unwrapping here because the RCC server should not return
+        // NoSuchPeripheral for a valid member of the Peripheral enum.
+        self.enter_reset_raw(peripheral as u32).unwrap()
+    }
+
+    /// Requests that the reset line to a peripheral be deasserted.
+    ///
+    /// This operation is idempotent and will be retried automatically should
+    /// the RCC server crash while processing it.
+    ///
+    /// # Panics
+    ///
+    /// If the RCC server has died.
+    pub fn leave_reset(&self, peripheral: Peripheral) {
+        // We are unwrapping here because the RCC server should not return
+        // NoSuchPeripheral for a valid member of the Peripheral enum.
+        self.leave_reset_raw(peripheral as u32).unwrap()
+    }
+}
+
+//
+// A few macros for purposes of defining the Peripheral enum in terms that our
+// driver is expecting:
+//
+// - RCC_IOPENR[31:0] and RCC_IOPRSTR[31:0] are indices 31-0.
+// - RCC_AHBENR[31:0] and RCC_AHBRSTR[31:0] are indices 63-32.
+// - RCC_APBENR1[31:0] and RCC_APBRSTR1[31:0] are indices 95-64.
+// - RCC_APBENR2[31:0] and RCC_APBRSTR2[31:0] are indices 127-96.
+//
+macro_rules! iop {
```

Comment:
What about using a `const fn` here instead of a macro?

(it looks like this is also applicable to the `stm32h7-rcc-api`!)

Reply:
I believe that, at the time the original code (which was for the f4) was written, const fns were not permitted in this syntactic position (the definition of enum values)? But they definitely are now! Will fix this one. Would you be willing to file a brief issue to ensure that we go get the f4/h7/etc?

Reply:
Done, opened #404 [PR NOT FOUND]